streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -751,75 +751,91 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
-          ZChannel.unwrap[Env1, Any, Any, Any, OutErr1, OutElem2, OutDone] {
-            outgoing.take.flatMap {
-              case s: Exit.Success[OutElem2] => Exit.succeed(ZChannel.write(s.value) *> writer)
-              case f: Exit.Failure[Either[Unit, OutDone]] =>
-                val failure0 = failure.unsafe.get(Unsafe)
-                val out = f.cause.failureOrCause match {
-                  case Left(_: Left[Unit, OutDone]) => ZChannel.refailCause(failure0)
-                  case Left(x: Right[Unit, OutDone]) =>
-                    if (failure0 eq Cause.empty) ZChannel.succeedNow(x.value)
+      ZIO.fiberIdWith { fiberId =>
+        val input       = SingleProducerAsyncInput.unsafe.make[InErr, InElem, InDone](fiberId)(Unsafe)
+        val queueReader = ZChannel.fromInput(input)
+        val n0          = n.toLong
+        val bufferSize0 = bufferSize
+        val outgoing    = Queue.unsafe.bounded[ResultUnordered](bufferSize0, fiberId)(Unsafe)
+        val errorSignal = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe)
+        val permits     = Semaphore.unsafe.make(n0)(Unsafe)
+        val failure     = Ref.unsafe.make[Cause[OutErr1]](Cause.empty)(Unsafe)
+
+        val setFinalizer: UIO[Unit] =
+          ZIO.uninterruptible {
+            scope.addFinalizer(outgoing.shutdown)
+          }
+
+        for {
+          _          <- setFinalizer
+          pull       <- (queueReader >>> self).toPullInAlt(scope)
+          childScope <- scope.fork
+          processElems = pull.flatMap { outElem =>
+                           val latch = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe)
+
+                           permits
+                             .withPermit(
+                               latch.succeedUnit *>
+                                 f(outElem)
+                                   .foldCauseZIO(
+                                     cause =>
+                                       failure.update(_ && cause).unless(cause.isInterruptedOnly) *>
+                                         errorSignal.succeedUnit *>
+                                         outgoing.offer(ResultUnordered.Failure),
+                                     elem => outgoing.offer(ResultUnordered.Value(elem))
+                                   )
+                             )
+                             .interruptible
+                             .forkIn(childScope) *> latch.await
+                         }.forever.interruptible
+                           .onError(_.failureOrCause match {
+                             case Left(x: Left[OutErr, OutDone]) =>
+                               failure.update(_ && Cause.fail(x.value)) *>
+                                 outgoing.offer(ResultUnordered.Failure)
+                             case Left(x: Right[OutErr, OutDone]) =>
+                               permits.withPermits(n0)(ZIO.unit).interruptible *>
+                                 outgoing.offer(ResultUnordered.Done(x.value))
```

Comment:
This would previously offer an `Exit.Failure(Right(...))` but now it offers a `Done` instead of `Failure`. Is that intended?